### ISSUE TITLE
Merge LH5 files from multiple threads

### ISSUE THREAD

**Issue by gipert** [legend-exp/remage#190]
Is it possible to handle this in pure C++ in a reasonable way? If not, we should provide some Python remage wrapper that performs the concatenation.

**Comment by Yurivanderburg** [comment]
Not sure if this helps, but I found loading the lh5 files as awkward arrays and use ak.concatenate works very nicely. I can share some code if you want.

**Comment by tdixon97** [comment]
This is exactly the approach we suggest people to use. 
However, the combination of the lh5 files directly in C++ / with an extra python script is a bit more tricky since you do not want to just concatenate but also sort by `g4_evtid` (since the files each contain a random set of `g4_evtids`). Then this also has to be done in a memory efficiency way (not reading the full data into memory), and it should be fast.

**Comment by gipert** [comment]
We also want to do some merging in the simulation production workflow to avoid cluttering the filesystem with a huge amount of files.

**Comment by gipert** [comment]
Now technically possible with #210.

**Comment by ManuelHu** [comment]
Do we want to try to guess the output file names ourselves in python code, or do we want to have some way to communicate data from the `remage-cpp` process to the python wrapper?

**Comment by gipert** [comment]
>the easies (infrastructure-wise) option would be to replicate the output file reanming scheme in python...

this is the way...

>there is also another problem: the number of threads also can be (indirectly) influenced via various env variables

then we'll have to understand how this works and implement protections in the python wrapper...

**Comment by ManuelHu** [comment]
another complication: the outpur file name can be set in the macro (and not in the command line args). In this case we need remage reporting back the details to the python wrapper _anyway_ (unless we want to re-implement macro parsing in python ... 🫣)

**Comment by tdixon97** [comment]
If remage now has a python wrapper we could potentially move away from geant4 macro files? Which are horrible.

**Comment by gipert** [comment]
I am not sure we need to support naming output files with macro commands. Maybe we can just remove the functionality.

Toby, that sounds complicated. You still need a way to pass the configuration options to remage-cpp.

Manuel, should we think about moving all the code you wrote to make LH5 files to the Python wrapper at some point?

**Comment by ManuelHu** [comment]
> should we think about moving all the code you wrote to make LH5 files to the Python wrapper at some point?

maybe, there were problems with h5py and the geant4-written files, the last time I tried.

Don't get me wrong, it would be very nice to get rid of this C++ code.

But if we do so, the same problem ("how to propagate parameters/data") would also exist for input files. There we do not even have command line switches (and in the current architecture we cannot inject data from the command line into generators, because at that point it is not really clear what generator the user will actually choose...) We might even have multiple different input files for different processes/generators/vertex generators...

**Comment by ManuelHu** [comment]
another complication: files _might_ not be there. Geant4 deletes empty output files, so they might be missing. (of course, they also don't tell you that they deleted the output file that was requested)

**Comment by tdixon97** [comment]
@gipert we need to revive this

**Comment by ManuelHu** [comment]
on the remage side everything is prepared, only changes for lgdo are missing (adding it to remage would be < 10 lines of code then)

**Comment by tdixon97** [comment]
ok ill have a look at modifying lh5concat

**Comment by ManuelHu** [comment]
two things have to be done:
* move the logic of `lh5concat` out of the cli module into some more reasonable place, where it can be called from python code. This should be straighforward
* make lh5concat not read whole objects in memory (i.e. use LH5Iterator where applicable)

**Comment by tdixon97** [comment]
Have a look here https://github.com/legend-exp/legend-pydataobj/pull/137

**Comment by tdixon97** [comment]
What is needed on the remage side @ManuelHu ?

**Comment by ManuelHu** [comment]
I already have a sketch of it here on my machine. Let me just tidy it up and push (probably tomorrow)

**Comment by EricMEsch** [comment]
Was this completed with #314? If not just re-open it i guess.

**Comment by ManuelHu** [comment]
no, #314 was just infrastructure - no actual merging happens yet (this is also tightly related to our discussion lately, if we want to do some simple post-processing in the wrapper, we just might skip the merging and let reboost do it
